File: Swift/ExampleTests/common/ResultValueString.swift
public struct ResultValueString: CustomStringConvertible, Equatable {
    public let sum: String

    public init(sum: String) {
        self.sum = sum
    }

    public init(fromArray v: [String]) {
        self.sum = v.count > 0 ? v[0] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> ResultValueString {
        let parts = Tokens.require(text, 1, "ResultValue")
        return ResultValueString(sum: parts[0])
    }

    public var description: String {
        return Tokens.token(sum)
    }

    public static let dncString = "?DNC?"

    public static func == (a: ResultValueString, b: ResultValueString) -> Bool {
        return (a.sum == dncString || b.sum == dncString || a.sum == b.sum)
    }
}

File: JavaScript/ExampleTests/common/ResultValueString.js
import * as tokens from "./tokens.js";

export class ResultValueString {
  static DNC_STRING = "?DNC?";

  constructor(sum = "") {
    this.sum = sum;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new ResultValueString(
      v[0] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 1, "ResultValue");
    return new ResultValueString(
      parts[0]
    );
  }

  toString() {
    return tokens.token(this.sum);
  }

  equals(other) {
    if (!(other instanceof ResultValueString)) return false;
    return (this.sum === ResultValueString.DNC_STRING || other.sum === ResultValueString.DNC_STRING || this.sum === other.sum);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}
